Fobnail api (#40)
fobnail/docs: docs/fobnail-api.md

### Fobnail Data format

Fobnail uses CBOR (with a few exceptions) for transmitting structured data both
from Fobnail to Attester and from Attester to Fobnail. We use subset of
[RFC7049](https://datatracker.ietf.org/doc/html/rfc7049) (no tagging and no
extensions).

Clients are **required** to set `Content-Format` to `Application/CBOR` when
transmitting CBOR-encoded data. Clients **may** set `Accept` header, in that
case server may return an error as defined
[here](https://datatracker.ietf.org/doc/html/rfc7252#section-5.10.4).

## API endpoints

### Fobnail Token provisioning

| Endpoint Name             | Method | Arguments                               |
| ------------------------- | ------ | --------------------------------------- |
| /admin/token_provision    | POST   | PO certificate chain                    |
| /admin/provision_complete | POST   | Fobnail Identity/Encryption certificate |

These APIs may be executed only when Fobnail Token is in a unprovisioned state.
Fobnail Token **must** respond with **4.03** if these APIs are called after
provisioning is complete.

#### /admin/token_provision

Client transfers PO certificate chain in the same format as used for EK

> certificate chain. Fobnail Token verifies certificate chain against the trust

Reviewer: EK certificate chain was not described yet. In order to reduce jumping over the document I would move description of format here and link to it in EK.

Author: Fixed

                // the same size, if these don't match then entire RIM is
                // considered invalid
                [ 0x00, 0x00, 0x00, 0x00, ... ],
                ...
            ]
        },
        // Another PCR bank (major 5, map)
        {
            "algo_id": 0xb,
            "pcrs": 0x000000ff,
            // major 4 (array)
            "pcr": [
                // major 2 (byte string)
                [ 0x00, 0x00, 0x00, 0x00, ... ],
                ...
            ]
        }
    ],
}
```

Client **must** sign RIM with AIK (see [Data signing](#data-signing) section
below).

Fobnail Token verifies RIM signature and format, if signature verification fails
**4.03** is returned, if RIM has invalid format **4.00** is returned. On success
**2.01** or **2.04** is returned. On the first successful call **2.01** is
returned (`Location-Path` must not be present), on subsequent calls **2.04** is
returned to signify that the previous RIM has been replaced with the new RIM.

#### /admin/provision/{id}

POST request to this endpoint results in writing data attached to the
Provisioning Context into persistent storage. Currently, this request does not

> contain any data and payload **must** be empty and or **4.00** response will be

Reviewer: ```suggestion
contain any data and payload **must** be empty or **4.00** response will be
```

Also, do we expect any payload in the future? i think if there were anything to be passed to Token, it should have a dedicated API.

Author: Fixed

# Fobnail API

Fobnail Token exposes API over
[CoAP](https://datatracker.ietf.org/doc/html/rfc7252) server exposed over UDP.

## General considerations

> This document describes API (major) version 1, and all API endpoint addresses

Author: Added

are relative to (starting with) `/api/v1`. The version will be increased only
when backward-incompatible changes are done.

### General request handling rules

Some APIs create temporary objects that live at most as long as the client is
connected. To check connection status, the server issues CoAP Ping requests,
which clients **must** respond to in order to avoid losing their objects.

When an object is created, the server responds with **2.01**, `Location-Path`
specifying the resource's path (a numeric object ID). Object IDs are specific to
a client - a client cannot access object IDs of other clients. `Location-Path`
is a string option, whereas object ID is an integer, so it must be encoded as a
string. Object IDs may be passed as part of the CBOR payload (as an integer) or
in URI (as a string).

If the `Content-Format` option is not present, Fobnail Token assumes that
`Content-Format` is `application/octet-stream`. All requests which pass
CBOR-encoded data **must** set `Content-Format` option to `application/cbor` for
Fobnail Token to interpret the payload as CBOR. The server responds with

> **4.00** (Bad Request) if unknown or invalid' Content-Format' is given. If

Reviewer: ```suggestion
**4.00** (Bad Request) if unknown or invalid `Content-Format` is given. If
```

[Conditional Request Options](https://datatracker.ietf.org/doc/html/rfc7252#section-5.10.8)
are **not** supported and will result in **4.02** (Bad Option). The payload
should contain a human-readable message describing what options are not
supported. See
[RFC7252 section 5.4.1](https://datatracker.ietf.org/doc/html/rfc7252#section-5.4.1)
for details.

Each response from the Fobnail Token **must** contain the `Content-Format`
option unless it is an error response with a diagnostic payload. See
[RFC7252](https://datatracker.ietf.org/doc/html/rfc7252#section-5.5.1) for
details.

### Error responses

Fobnail Token may return error codes as defined by the CoAP specification. An
error response **may** contain an additional context sent as a payload. All
cacheable error responses **must** have `Max-Age` set to 0.

Error response must not contain the `Content-Format` option. Message (if present)
must be passed as a raw UTF-8 string.

### Fobnail Data format

Fobnail uses CBOR (with a few exceptions) to transmit structured data from
Fobnail to Attester and from Attester to Fobnail. We use a subset of
[RFC7049](https://datatracker.ietf.org/doc/html/rfc7049) (no tagging and no
extensions).

> Clients are **required** to set `Content-Format` to `Application/CBOR` when

Reviewer: ```suggestion
Clients are **required** to set `Content-Format` to `application/cbor` when
```

transmitting CBOR-encoded data.

### Data signing

A signature is generated by taking an already CBOR-encoded object, a nonce

> (generated by Fobnail and sent as part of the request), and signing the hash of

Reviewer: Fobnail doesn't generate requests anymore. Also, since this is not part of single request/response pair, nonce would probably have to be returned in next request from client, in case another client (or attacker) tries to connect in between those two requests (generate nonce and send signed data).

Author: @krystian-hebel We can't trust that the nonce we get from client is real. If attacker can replay signed message he may also provide replayed nonce. In some cases we could use nonce client-provided nonce - when Fobnail signs data and client verifies it but not when Fobnail verifies that data.

Reviewer: I'm not talking about generating nonce on platform, just not stripping it from next request. That way server doesn't have to check it against every generated nonce, in case there are multiple simultaneously open connections in the future.

Also, how attester is supposed to get nonce for `/attest`?

Author: Server keeps a single nonce for each client so it has to compare only with that nonce. I added /nonce endpoint for reading nonce.

both of them. The nonce is appended to the CBOR blob as plain bytes. The
resulting signature and original data (without nonce) are wrapped into another
CBOR object:

```json
{
    // The data we signed (nested CBOR, without nonce)
    // major 2 (byte array)
    "data": [0x42, ...],
    // The signature we just computed (major 2, byte array)
    "signature": [0x20, ...]
}
```

#### Exception

> Result of `quote` command has nonce explicitly included in returned structure.

Reviewer: ```suggestion
Result of `quote` command has nonce explicitly included in sent structure.
```
(unless this is reworded after change requested above)

format as above. This (along with other mechanisms) ensures freshness of Claims,
in addition to freshness of Evidence (see RATS architecture for description of
those artifacts).

## API endpoints

### Fobnail Token provisioning

| Endpoint Name             | Method | Arguments                               |
| ------------------------- | ------ | --------------------------------------- |
| /admin/token_provision    | POST   | PO certificate chain                    |
| /admin/provision_complete | POST   | Fobnail Identity/Encryption certificate |

These APIs may only be executed when Fobnail Token is in an unprovisioned state.
Fobnail Token **must** respond with **4.03** if these APIs are called after
provisioning is complete.

#### /admin/token_provision

PO certificate chain is transferred as an ordered array of raw DER certificates,
starting with the certificate directly after the root certificate (the Fobnail
Token already knows the root).

```json
{
    // Array of certificates in DER format (major 4)
    "certs": [
        // CA immediately under root (major 2)
        [ 0x30, 0x82, 0x03, 0x8b, ... ],
        // Intermediate CAs, if any
        ...

>         // EK certificate (major 2)

Reviewer: ```suggestion
        // PO certificate for certificate signing (major 2)
```

AIK object ID is invalid, **4.04** is returned.

#### /admin/provision/{id}/meta

This method is used to upload platform metadata. After upload, the platform
metadata is bound to the provisioning context specified by `{id}`.

Metadata is sent in CBOR-encoded format (also described
[here](https://github.com/fobnail/docs/blob/main/dev-notes/metadata.md) in more
detail). Example metadata looks like this

```json
{
    // metadata version, used to detect backward-incompatible changes
    // (major 0, unsigned integer)
    "version": 1,
    // device manufacturer extracted from SMBIOS tables (major 3, UTF-8 string)
    "manufacturer": "Gigabyte",
    // device model extracted from SMBIOS tables (major 3, UTF-8 string)
    "model": "Gigabyte A520 AORUS ELITE",
    // MAC address of primary network card (major 2, byte string)
    "mac": [0xea, 0x96, 0x91, 0x87, 0x26, 0x8d],
    // device serial number extracted from SMBIOS tables (major 3, UTF-8 string)
    "sn": "S21N559B431",
}
```

Client **must** sign metadata with AIK (see the [Data signing](#data-signing)
section above).

Fobnail Token verifies the provided metadata against AIK bound to the

> Provisioning Context, if signature verification fails, **4.03** is returned. If

Reviewer: ```suggestion
Provisioning Context. If signature verification fails, **4.03** is returned. If
```

        {
            "algo_id": 0x000b,
            "pcrs": 0x000000ff,
        }
    ],
    // Nonce to be passed to TPM_Quote() (major 2)
    "nonce": [0x00, 0x01, 0x02, 0x03, ...]
}
```

The PCR selection format is similar to that of [RIM](#adminprovisionidrim) but
stripped of unnecessary fields. Due to the nature of PCR selection parsing done
by `TPM2_Quote()`, order of PCR banks matters.

#### /attest/{id}

The client sends the evidence to this endpoint (the result of TPM Quote). Based
on the evidence, Fobnail Token decides whether the platform is trustworthy or
not. If Fobnail Token decides that the platform is trustworthy, **2.04**
response is returned, **4.03** otherwise.

If attestation is successful, Fobnail Token unlocks access to the Fobnail Token
Services.

### Fobnail Token Services

Fobnail Token Services are available after successful platform attestation.
Fobnail Token can store cryptographic keys (or other stuff) in its internal
flash. Depending on the key's type (symmetric or asymmetric) and usage
permissions, various operations are available. For symmetric keys, it is
encryption and decryption. For asymmetric keys, it is signing, decryption, KDF,

> and public key read (The client extracts the public key and does encryption on

Reviewer: ```suggestion
and public key read (the client extracts the public key and does encryption on
```

| /storage/fs/{name}   | PUT    | File name and file contents      |
| /storage/fs/{name}   | DELETE | File name                        |

If the platform has not been attested (or failed attestation), these APIs will
return a **4.04** error. If attestation is successful, Fobnail Token checks
whether a specified key exists and whether the platform has access to that key.
If the key or file does not exist or there is no access **4.04** error is
returned.

> Note: currently, crypto API is not fully defined. Future versions of this
> document will address this problem.

#### GET /storage/fs/{name}

The client sends a request to this endpoint to read a file. The file path is
encoded as part of the URI. If the file exists (and if it's accessible),
**2.05** response is sent with the file contents as the payload. `Max-Age`
option **must** be set to 0 to prevent caching. ETag option **must not** be
present in the response, and if present in the request, it **must** be ignored.

> The GET method always returns the full contents of the file. Further versions
> of this document may define API using the FETCH method from RFC8132.

> Fobnail Token attempts to open (or create) the requested file and returns

Reviewer: Why create?

Author: Forgot to change that paragraph; previous version required to open the file before reading but we don't need to keep any context here as we do for some crypto operations so I simplified API.

Author: Fixed

Error response must not contain the `Content-Format` option. Message (if present)
must be passed as a raw UTF-8 string.

### Fobnail Data format

Fobnail uses CBOR (with a few exceptions) to transmit structured data from
Fobnail to Attester and from Attester to Fobnail. We use a subset of
[RFC7049](https://datatracker.ietf.org/doc/html/rfc7049) (no tagging and no
extensions).

Clients are **required** to set `Content-Format` to `application/cbor` when
transmitting CBOR-encoded data.

### Data signing

A signature is generated by taking an already CBOR-encoded object, a nonce
(generated by Fobnail), and signing the hash of both of them. The nonce is
appended to the CBOR blob as plain bytes. The resulting signature and original
data (without nonce) are wrapped into another CBOR object:

```json
{
    // The data we signed (nested CBOR, without nonce)
    // major 2 (byte array)
    "data": [0x42, ...],
    // The signature we just computed (major 2, byte array)
    "signature": [0x20, ...]
}
```

To obtain nonce for data signing, please use the [/nonce](#nonce) endpoint.

#### Exception

Result of `quote` command has nonce explicitly included in sent structure.

> No additional data is appended before signing. Returned CBOR object has the same

Reviewer: ```suggestion
No additional data is appended before signing. Sent CBOR object has the same
```

subsequent requests to the Provisioning Context will return **4.04**.

### Attestation

| Endpoint Name             | Method | Arguments            |
| ------------------------- | ------ | -------------------- |
| /attest                   | POST   | Platform metadata    |
| /attest/{id}              | POST   | Output of TPM quote  |

#### /attest

The client sends platform metadata in the same format as when calling
`/admin/provision/{id}/meta`. Metadata **must** be signed by the same AIK that
was used during platform provisioning. If verification is successful **2.01** is
returned (otherwise **4.04**) and `Location-Path` contains Attestation Context
ID. The response's payload contains PCR selection and nonce:

```json
{
    // Array of PCR banks (major 4)
    "banks": [
        // First PCR bank
        {
            // ID of algorithm used by PCRs (major 0)
            "algo_id": 0x0004,
            // Bitmap of present PCRs (major 0, unsigned integer)
            "pcrs": 0xffffffff,
        },
        // Another PCR bank (major 5, map)
        {
            "algo_id": 0x000b,
            "pcrs": 0x000000ff,
        }
    ],
    // Nonce to be passed to TPM_Quote() (major 2)

>     "nonce": [0x00, 0x01, 0x02, 0x03, ...]

Reviewer: This partially overlaps with new `/nonce` endpoint. I'm OK with keeping it in the structure to minimize number of requests, but in that case we need to strictly define whether `/nonce` is allowed between `/attest` and `/attest/{id}`.

I'd say it should be forbidden, and result in invalidation of attestation context ID, in preparation for another `/attest` message. Otherwise compromised client can keep requesting new `/nonce` until it receives a nonce that was used in previous quote and replay that message.

Author: Done

}
```

The PCR selection format is similar to that of [RIM](#adminprovisionidrim) but
stripped of unnecessary fields. Due to the nature of PCR selection parsing done
by `TPM2_Quote()`, order of PCR banks matters.

#### /attest/{id}

The client sends the evidence to this endpoint (the result of TPM Quote). Based
on the evidence, Fobnail Token decides whether the platform is trustworthy or
not. If Fobnail Token decides that the platform is trustworthy, **2.04**
response is returned, **4.03** otherwise.

If attestation is successful, Fobnail Token unlocks access to the Fobnail Token
Services.

### Fobnail Token Services

Fobnail Token Services are available after successful platform attestation.

> Fobnail Token can store cryptographic keys (or other stuff) in its internal

Reviewer: ```suggestion
Fobnail Token can store cryptographic keys (or other data) in its internal
```
